stackdriver/config/config.go

```diff
@@ -0,0 +1,168 @@
+
+	// ProjectIDKey is the map key for Config.ProjectID.
+	ProjectIDKey = "project-id"
+	// GcpLocationKey is the map key for Config.GCPLocation.
+	GcpLocationKey = "gcp-location"
+	// ClusterNameKey is the map key for Config.ClusterName.
+	ClusterNameKey = "cluster-name"
+	// GcpSecretNameKey is the map key for GCPSecretName.
+	GcpSecretNameKey = "gcp-secret-name"
+	// GcpSecretNamespaceKey is the map key for GCPSecretNamespace.
+	GcpSecretNamespaceKey = "gcp-secret-namespace"
+)
+
+// Config encapsulates the metadata required to configure a Stackdriver client.
+type Config struct {
+	// ProjectID is the stackdriver project ID to which data is uploaded.
+	// This is not necessarily the GCP project ID where the Kubernetes cluster is hosted.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	ProjectID string
+	// GCPLocation is the GCP region or zone to which data is uploaded.
+	// This is not necessarily the GCP location where the Kubernetes cluster is hosted.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	GCPLocation string
+	// ClusterName is the cluster name with which the data will be associated in Stackdriver.
+	// Required when the Kubernetes cluster is not hosted on GCE.
+	ClusterName string
+	// GCPSecretName is the optional GCP service account key which will be used to
+	// authenticate with Stackdriver. If not provided, Google Application Default Credentials
+	// will be used (https://cloud.google.com/docs/authentication/production).
+	GCPSecretName string
+	// GCPSecretNamespace is the Kubernetes namespace where GcpSecretName is located.
+	// The Kubernetes ServiceAccount used by the pod that is exporting data to
+	// Stackdriver should have access to Secrets in this namespace.
+	GCPSecretNamespace string
+}
+
+// NewStackdriverConfigFromConfigMap returns a Config for the given configmap
+func NewStackdriverConfigFromConfigMap(config *corev1.ConfigMap) (*Config, error) {
+	if config == nil {
+		return &Config{}, nil
+	}
+	return NewStackdriverConfigFromMap(config.Data)
+}
+
+// NewStackdriverConfigFromMap returns a Config for the given map
+func NewStackdriverConfigFromMap(config map[string]string) (*Config, error) {
+	sc := &Config{}
+
+	if pi, ok := config[ProjectIDKey]; ok {
+		sc.ProjectID = pi
+	}
+
+	if gl, ok := config[GcpLocationKey]; ok {
+		sc.GCPLocation = gl
+	}
+
+	if cn, ok := config[ClusterNameKey]; ok {
+		sc.ClusterName = cn
+	}
+
+	if gsn, ok := config[GcpSecretNameKey]; ok {
+		sc.GCPSecretName = gsn
+	}
+
+	if gsns, ok := config[GcpSecretNamespaceKey]; ok {
+		sc.GCPSecretNamespace = gsns
+	}
+
+	return sc, nil
+}
+
+// GetStackdriverSecret returns the Kubernetes Secret specified in the given config.
+func GetStackdriverSecret(config *Config) (*v1.Secret, error) {
+	ensureKubeclient()
+	ns := secretNamespaceDefault
+	if config.GCPSecretNamespace != "" {
+		ns = config.GCPSecretNamespace
+	}
+
+	return kubeclient.CoreV1().Secrets(ns).Get(config.GCPSecretName, metav1.GetOptions{})
+}
```

Comment:
I am wondering if this should be part of Config.
This probably should be done where the config is used and let Config logic be simple parsing of the ConfigMap.

Reply:
I would say since "stackdriver/config" defines the definition of GCPSecret being a Kubernetes secret name, it should also provide the logic of how to use it. 

It also prevents a consumers that don't need a kubeclient for other purposes than reading this Secret from having to re-implement the kubeclient bit.